### PR TITLE
[FLINK-6107] Add custom checkstyle for flink-streaming-java

### PR DESCRIPTION
**Pull request by aljoscha**
I played around with this a bit over the weekend... 😃 

This is a proposal for a way forward from the discussion on code style on the mailing list: https://lists.apache.org/thread.html/94c8c5186b315c58c3f8aaf536501b99e8b92ee97b0034dee295ff6a@%3Cdev.flink.apache.org%3E.

This adds a custom `checkstyle.xml` based on the Apache Beam checkstyle for `flink-streaming-java` only. The code style is not really different form what we currently have, this just aims at unifying the existing codebase. The Flink specific changes are mostly:
 - Tabs instead of spaces
 - No line-length limit
 - We disallow a bunch of imports

The first commit adds the checkstyle file but a lot of stuff is commented out because the Flink code base would fail those tests. The following commits subsequently add more and more invasive changes along with enabling more rules.

The commits up to the last two commits at uncontroversial (in my opinion) checks that will make code reviews a lot easier in the future. Especially having an enforced import order and trailing whitespace policy will make import order changes and whitespace changes a thing of the past. The checks for spacing around unary and binary operators should also be unproblematic.

The last two commits normalise the style of code blocks. Having a unified style for those should be good in the future and will prevent "edit wars" (same as for import order and whitespace) with people that have their IDE formatter configured differently.

### CODE REVIEW

**Comment by zentol** [comment]
Love the checks that were enabled. They really cover most (if not all) issues that we have to raise in reviews at the moment.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/collector/selector/OutputSelectorWrapper.java [hunk 1]
```diff
@@ -1,28 +0,0 @@
...
- */
-
-package org.apache.flink.streaming.api.collector.selector;
-
-import java.io.Serializable;
-
-import org.apache.flink.annotation.PublicEvolving;
-
-@PublicEvolving
-public interface OutputSelectorWrapper<OUT> extends Serializable {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/collector/selector/OutputSelectorWrapper.java]: unused class?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/ReduceIterableAllWindowFunction.java [hunk 2]
```diff
@@ -1,48 +0,0 @@
...
- */
-package org.apache.flink.streaming.api.functions.windowing;
-
-import org.apache.flink.annotation.Internal;
-import org.apache.flink.api.common.functions.ReduceFunction;
-import org.apache.flink.streaming.api.windowing.windows.Window;
-import org.apache.flink.util.Collector;
-
-@Internal
-public class ReduceIterableAllWindowFunction<W extends Window, T> implements AllWindowFunction<T, T, W> {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/ReduceIterableAllWindowFunction.java]: Unused class?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/ReduceIterableWindowFunction.java [hunk 3]
```diff
@@ -1,48 +0,0 @@
...
- */
-package org.apache.flink.streaming.api.functions.windowing;
-
-import org.apache.flink.annotation.Internal;
-import org.apache.flink.api.common.functions.ReduceFunction;
-import org.apache.flink.streaming.api.windowing.windows.Window;
-import org.apache.flink.util.Collector;
-
-@Internal
-public class ReduceIterableWindowFunction<K, W extends Window, T> implements WindowFunction<T, T, K, W> {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/ReduceIterableWindowFunction.java]: Unused class?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/delta/extractor/ArrayFromTuple.java [hunk 4]
```diff
@@ -28,7 +28,7 @@
 public class ArrayFromTuple implements Extractor<Tuple, Object[]> {
 
 	/**
-	 * Auto generated version id
+	 * Auto generated version id.
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/delta/extractor/ArrayFromTuple.java]: let's remove this like in ExtractionAwareDeltaFunction

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamingReader.java [hunk 5]
```diff
@@ -1,30 +0,0 @@
...
-
-package org.apache.flink.streaming.runtime.io;
-
-import org.apache.flink.annotation.Internal;
-import org.apache.flink.runtime.io.network.api.reader.ReaderBase;
-
-import java.io.IOException;
-
-@Internal
-public interface StreamingReader extends ReaderBase {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamingReader.java]: Unused class?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/delta/extractor/ArrayFromTuple.java (hunk 4 above)
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/windowing/delta/extractor/ArrayFromTuple.java]: dammit, I missed this one...

Fixing

**Comment by aljoscha** [comment]
Thanks for reviewing! 😃 

And yes, the classes I removed where unused.

**Comment by StephanEwen** [comment]
Can you describe what rules this style actually enforces?

**Comment by zentol** [comment]
@StephanEwen It enforeces the following rules:
* every files has to end with a new-line
* no trailing whitespace *anywhere*
* every public/protected method/class must have a javadoc
* imports have to be sorted alphabetically
* static imports must come first
* whitespace after commas, semicolons and casts
* whitespace around operators, if, for, etc.
* left curly brace (`{`) must not be at the start of the line
```
public MyObject(int a)
{

=>

public MyObject(int a) {
```
* else/catch/finally must be on the same line as the right curly brace (`}`)
```
if {
}
else

=>

if {
} else
```
* static final variables must be upper case
* non-static variables must not be upper case
```
```

**Comment by StephanEwen** [comment]
I like to the idea to start this with `flink-streaming-java`, but do we have a discussion somewhere about the code style we want to enforce? My gut feeling is that we don't quite yet agree on all details.

**Comment by aljoscha** [comment]
In the ML discussion "Google code style" was repeatedly brought up (initially by me): https://lists.apache.org/thread.html/94c8c5186b315c58c3f8aaf536501b99e8b92ee97b0034dee295ff6a@%3Cdev.flink.apache.org%3E.

I posted this proposal to encourage discussion, that's why I also wrote to the ML again. I think all checks except `RightCurly` and `LeftCurly` are pretty standard Java style so the discussion will probably revolve around those. Or we decide to not enforce a style for those.

**Comment by greghogan** [comment]
@aljoscha do you think we can use a single checkstyle or will this need to be customized per module? Is this enforced by IntelliJ?

**Comment by aljoscha** [comment]
We will need at least two: the existing checkstyle for all the other modules and the new strict style. (We might get away with having one checkstyle and per-module suppression files that deactivate rules but I think this can lead to more hassle in the end.)

IntelliJ can be setup to use a custom checkstyle per module. In my setup all the modules use the default Flink checkstyle, only for flink-streaming-java uses the new strict style to mirror the maven setup.

**Comment by greghogan** [comment]
@aljoscha should we host the new checkstyle under tools/maven/ alongside the existing checkstyle? There is already a ticket (FLINK-6137) to add a custom checkstyle to flink-cep and I don't see any of these rules being module-specific.

**Comment by StephanEwen** [comment]
It would be awesome if we could make IntelliJ pick up the style config
automatically. By  committing some files under ".Idea", we might be able to
do that...

**Comment by dawidwys** [comment]
+1 for hosting the checkstyle under tools/maven/

**Comment by aljoscha** [comment]
AFAIK it's not possible to make IntelliJ pick up the checkstyle config by itself. On my machine the setting is in the "*.iml" project file and I think we have so far not included IDE specific project files because this is quite uncommon, I think.

@dawidwys I'll create a new maven module to host the maven-build related files. This will make it easier to include it in different modules without dealing with paths in the checkstyle config.

**Comment by zentol** [comment]
Could you create or link a guide on how to configure checkstyles per module in Intellij?

**Comment by aljoscha** [comment]
@zentol I added a section to our IDE setup Guide.

File: tools/maven/strict-checkstyle.xml [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+      <property name="message"
+                value="Line has leading space characters; indentation should be performed with tabs only." />
+      <property name="ignoreComments" value="true" />
+    </module>
+
+    <!-- Prohibit T.getT() methods for standard boxed types -->
+    <module name="Regexp">
+      <property name="format" value="Boolean\.getBoolean"/>
+      <property name="illegalPattern" value="true"/>
+      <property name="message" value="Use System.getProperties() to get system properties."/>
```
Comment (aljoscha) [inline tools/maven/strict-checkstyle.xml]: This is the Javadoc for the method that the checks are referring to:
```
/**
 * Returns {@code true} if and only if the system property
 * named by the argument exists and is equal to the string
 * {@code "true"}. (Beginning with version 1.0.2 of the
 * Java<small><sup>TM</sup></small> platform, the test of
 * this string is case insensitive.) A system property is accessible
 * through {@code getProperty}, a method defined by the
 * {@code System} class.
 * <p>
 * If there is no property with the specified name, or if the specified
 * name is empty or null, then {@code false} is returned.
 *
 * @param   name   the system property name.
 * @return  the {@code boolean} value of the system property.
 * @throws  SecurityException for the same reasons as
 *          {@link System#getProperty(String) System.getProperty}
 * @see     java.lang.System#getProperty(java.lang.String)
 * @see     java.lang.System#getProperty(java.lang.String, java.lang.String)
 */
public static boolean getBoolean(String name) {
```

**Comment by aljoscha** [comment]
I have no preference for any style of import order. I just wanted to mandate some order so that we don't have edit wars when people use different IDE settings.

@greghogan Have you tried setting up import check settings that more closely match the current flink styl? If we find something that works I'm happy to change that. 😃

File: tools/maven/strict-checkstyle.xml (hunk 6 above)
Comment (greghogan) [inline tools/maven/strict-checkstyle.xml]: The section comment is `<!-- Prohibit T.getT() methods for standard boxed types -->`.

**Comment by greghogan** [comment]
@aljoscha, I likewise have no great preference for import order. I do think it is important for the checkstyle to match IntelliJ's code style, either the default or a provided Flink style.

The default IntelliJ import style can be approximated:

```
<module name="CustomImportOrder">
    <property name="customImportOrderRules" value="THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
    <property name="specialImportsRegExp" value="^javax\."/>
    <property name="standardPackageRegExp" value="^java\."/>
    <property name="sortImportsInGroupAlphabetically" value="true"/>
    <property name="separateLineBetweenGroups" value="false"/>
</module>
```

That includes a blank line between `javax` and `java` imports. There is an [old ticket with recent activity](https://github.com/checkstyle/checkstyle/issues/525) for this issue to allow blank lines to be explicitly defined.

That said, I'd go for the Google Style as used in this PR. IntelliJ can import from a checkstyle configuration but I am not seeing any effect from this.

File: tools/maven/strict-checkstyle.xml (hunk 6 above)
Comment (aljoscha) [inline tools/maven/strict-checkstyle.xml]: Ah, I should have mentioned that the Javadoc is for `Boolean.getBoolean()`. I copied these checks from our existing Flink checkstyle and I think the reason for prohibiting these is that it's very easy to confuse `Boolean.getBoolean(String)`, which reads from system properties, and `Boolean.parseBoolean(String)`, which parses the given string as a boolean. That's the reason behind the somewhat strange message which tells you to use `System.getProperties()` instead.

**Comment by aljoscha** [comment]
@greghogan Now I'm confused. 😅 Are you suggesting we use the Google style or switch to the default IntelliJ style?

**Comment by greghogan** [comment]
Google's style seems sensible since IntelliJ's style cannot be modeled in Checkstyle, the import listing is folded by default, and developers will need to load other non-default configuration settings into IntelliJ.

**Comment by zentol** [comment]
@aljoscha i believe enough earth rotations have passed :)

**Pull request merged**